### PR TITLE
decomp: finish `battlecontroller`

### PR DESCRIPTION
**Pull request by xTVaser**

### CODE REVIEW

File: decompiler/config/all-types.gc [hunk 1]
```diff
@@ -24636,20 +24639,20 @@
...
+   (enabled       symbol   :offset-assert 28)
    )
   :method-count-assert 9
   :size-assert         #x20
   :flag-assert         #x900000020
   )
 
 (deftype battlecontroller-creature-type (structure)
-  ((type2            basic  :offset-assert 0)
+  ((type2            nav-enemy  :offset-assert 0) ; a guess
```
Comment (water111) [inline decompiler/config/all-types.gc:24655]: probably a `type`

File: goal_src/levels/common/battlecontroller.gc [hunk 2]
```diff
@@ -5,3 +5,1043 @@
...
+   (spawn-count                 int16                                    :offset-assert 198)
+   (die-count                   int16                                    :offset-assert 200)
+   (target-count                int8                                     :offset-assert 202)
+   (spawner-count               int8                                     :offset-assert 203)
+   (creature-type-count         int8                                     :offset-assert 204)
+   (spawner-array               battlecontroller-spawner       8 :inline :offset-assert 208)
+   (spawn-period                uint64                                   :offset-assert 464)
+   (path-spawn                  path-control                             :offset-assert 472)
+   (creature-type-array         battlecontroller-creature-type 4 :inline :offset-assert 476)
+   (final-pickup-type           int32                                    :offset-assert 604)
```
Comment (water111) [inline goal_src/levels/common/battlecontroller.gc:51]: this is probably a `pickup-type`

File: goal_src/levels/common/battlecontroller.gc [hunk 3]
```diff
@@ -5,3 +5,1043 @@
...
+    (when
+     (and s4-0 (logtest? (-> (the-as nav-enemy s4-0) nav-enemy-flags) 2048))
+     (cond
+      ((< (-> s5-0 state) (-> s5-0 path curve num-cverts))
+       (when (or (-> self noticed-player) (= (-> s5-0 state) 1))
+        (let ((s3-0 (new 'stack-no-clear 'vector)))
+         (eval-path-curve-div!
+          (-> s5-0 path)
+          s3-0
+          (the float (-> s5-0 state))
```
Comment (water111) [inline goal_src/levels/common/battlecontroller.gc:194]: `eval-path-curve-div!` should probably take an `int` for this argument, but we don't have to worry about that now

File: goal_src/levels/common/battlecontroller.gc [hunk 4]
```diff
@@ -5,3 +5,1043 @@
...
+  0
+  (none)
+  )
+
+(defbehavior battlecontroller-battle-end battlecontroller ()
+  (case (-> (level-get-target-inside *level*) name) 
+   (('citadel)
+     (level-hint-spawn
+      (game-text-id misty-battle-finished)
+      "sksp0378"
```
Comment (water111) [inline goal_src/levels/common/battlecontroller.gc:499]: do you know what this says?

File: goal_src/levels/common/battlecontroller.gc [hunk 5]
```diff
@@ -5,3 +5,1043 @@
...
+    obj
+    arg0
+    )
+   )
+  )
+
+(defmethod deactivate battlecontroller ((obj battlecontroller))
+  (with-pp
+   (let ((gp-0 pp))
+    (let ((s6-1 obj))
```
Comment (water111) [inline goal_src/levels/common/battlecontroller.gc:815]: this is changing the process pointer, which is something we don't handle correctly in the decompiler yet.  can you leave a note at the top of the file about this? https://github.com/water111/jak-project/issues/882

File: goal_src/levels/common/battlecontroller.gc (hunk 4 above)
Comment (xTVaser) [reply goal_src/levels/common/battlecontroller.gc:499]: "the doors open, lets get going"

File: goal_src/levels/common/battlecontroller.gc (hunk 3 above)
Comment (xTVaser) [reply goal_src/levels/common/battlecontroller.gc:194]: hmm are you sure, it seems to explicitly convert to a float?
```asm
    mtc1 f0, s5               ;; [ 20] (set! f0-0 (gpr->fpr s5-0)) [s5: int ] -> []
    cvt.s.w f0, f0            ;; [ 21] (set! f0-1 (i2f f0-0)) [] -> []
    mfc1 a2, f0               ;; [ 22] (set! a2-0 (fpr->gpr f0-1)) [] -> [a2: float ]
```

File: goal_src/levels/common/battlecontroller.gc (hunk 4 above)
Comment (water111) [reply goal_src/levels/common/battlecontroller.gc:493]: It's under `citadel` in the case statement - could this be for the room with all the annoying hopping enemies?

File: goal_src/levels/common/battlecontroller.gc (hunk 3 above)
Comment (water111) [reply goal_src/levels/common/battlecontroller.gc:191]: oops, you're right - it should definitely stay the way it is. I read it as `(the-as float` instead.

**Pull request merged**